Add expected conditions 1827

emanlove opened the pull request:
Some very early and very basic sketches of what this might become ..

emanlove committed:
Sketching out initial ideas for test suite on expected conditions

emanlove committed:
Added dynamic and delayed title change as initial test framework for
expected conditions

emanlove committed:
Merge branch 'master' into add-expected-conditions-1827

emanlove committed:
Initial (working) prototype of wait for.. keyword

Wanted to just throw out a sample test case, test application (page),
and a working keyword. Interesting this works at a very basic level. See
next paths for a few functionality like

- instead of snake_case create method for using "space case"
- add check for presence of method
- add polling delay, timeout
- ...?

emanlove committed:
Bumped number of library keywords by one

emanlove committed:
Moved previous expected condition tests out of current test suite

These were some initial sketches for the keywords and their usage. As
they are not working I have moved them out of the test suite.

emanlove committed:
Adding another sample tests and modifying keyword behavior

- Added 100 millisecond polling to wait. Seems like a good default. Will
  need to decide on how to let users set this.
- Copied dynamic_content.html making into seperate
  expected_conditions.html test page.
- Added another sample test case

emanlove committed:
Added sample test case where an expected timeout would occur

yuriverweij commented on src/SeleniumLibrary/keywords/expectedconditions.py:
```diff
@@ -0,0 +1,25 @@
+        wait = WebDriverWait(self.driver, 10, 0.1)
+        # import sys,pdb;pdb.Pdb(stdout=sys.__stdout__).set_trace()
+        c = getattr(EC, condition)
+        wait.until(c(*args))
```
I've been looking through the expectedconditions documentation.
Depending on the expectedcondition it seems either a bool, webelement or string is returned.
Do we want to add these return values?  (probably yes)
This also adds verification possibilities.

emanlove replied:
I lean towards Yes but I also see there are a lot of different return values and combinations. So I am not sure what that means for properly setting the return. I also wonder if it is possible to set the return values based upon which expected condition is called?

I want to recommend that we put this into the next release and get expected conditions out there. Thoughts?

yuriverweij replied:
I think for now the "wait for" is basically a boolean already, and can be used as such. You also show this in the tests. So for now it is ok, and maybe at a later stage look at what is needed?

emanlove replied:
Ok, so I think we could set the type for the wait for expected condition keyword as you are right. That is singular in focus and has a return type. I was concerned about setting the return type on the (underlying) expected condition which seems a bit problematic or complicated based upon which expected condition it was. This was something i didn't want to hold up a release on given the amount of time and my lack of seeing how this can be done with multiple different method search with their own return type.

Now noting that the wait is a type and can be added now, looking at the selenium code I am confused about the type [[L81](https://github.com/SeleniumHQ/selenium/blob/trunk/py/selenium/webdriver/support/wait.py#L81)]. It looks like a generic type "T" with no really specifics on it. I would have thought it would have been Boolean as well. I don't know  if this is generic return type because Selenium didn't have a need to specify itself. Or if i am just understand the  type hinting here. I will note there is the similar until_not method in the same file which has an addition type `Literal[True]` as well. Still this doesn't help me to understand this but at least it is different.

yuriverweij replied:
I'm not familiar enough with typing to say this with full certainty, but to me it seems like:
"until" waits until the response is no longer False or Falsy, so for conditions where a boolean is returned this is obvious. For responses with a WebElement or anything else it's basically the same; no response is still Falsy/False, but when a response is received it becomes True/Truthy so the "until" succeeds. No specific type needed, and the return value could be anything.

emanlove replied:
Going to leave this as is for now ..

yuriverweij commented on atest/acceptance/keywords/expected_conditions.robot:
```diff
@@ -0,0 +1,21 @@
+Wait For Expected Conditions One Argument
+    Title Should Be    Original
+    Click Element    link=delayed change title
+    Wait For Expected Condition    title_is    Delayed
```
~If we get a return value for this we can verify the keyword. Right now we're verifying the Title Change and the wait, not necessarily the expected condition itself.  (The wait will wait for max timeout, after which the change will be done regardless if the expected condition returns true or false)~

Edit:
After doing some testing I see now that the wait fails if expected condition is not met. Still, returning the result might be useful since when it succeeds a WebElement/List/ etc can be returned

emanlove replied:
I see the advantage here. Not sure of a structure to do this is in. Again I would say let's do this in a follow up release if you ok with that.

yuriverweij commented on src/SeleniumLibrary/keywords/expectedconditions.py:
```diff
@@ -0,0 +1,25 @@
+
+class ExpectedConditionKeywords(LibraryComponent):
+    @keyword
+    def wait_for_expected_condition(self, condition, *args):
```
some brainstorming:
wait_for_expected_condition should take the following arguments:
- condition (condition=)
- timeout  (timeout=)
- msg (msg=custom message)
- arguments  (arguments for the condition, depending on the condition this can be a String, WebElement, List)

One (to me) "strange" expectedcondition is "new_window_is_opened". Does this one implicitly add the current number of windows?

emanlove replied:
Timeout has been added. Msg has not and that would be good to get in there. Let's make a separate ticket for that.

yuriverweij committed:
added timeout argument to keyword wait_for_expected_condition
Added timeout atest

emanlove committed:
Merge pull request #3 from yuriverweij/feature/add_expected_conditions_1827

added timeout argument to keyword wait_for_expected_condition

emanlove committed:
Some intial parsing and validation for parsing condition argument

emanlove committed:
Basic working keyword for wait

Made a few changes ..
- Catch unknown expoected conditions and reopport an error. This can still be improved
  by adding a here are the known expected conditions as well as a "did you mean ..."
  type message
- Brought in parser so one can use `Element To Be Clickable` instead of
  `element_to_be_clicable`. Admittedly it seems to simple and realized an issue, of case,
  as I type this.
- Added some test cases

emanlove committed:
Added test case for capitals within expected condition name

Hadn't realized I do force the expected condition ame to all lower case letters. So just needed
to add a test case!

emanlove committed:
Added example test where expected condition expected "locator" tuple

Some of the expected conditions expect a locator which is a tuple
consisting of the By string [1] and the string containing the locator.
I wanted to see what this was like. I am thinking that it would be good
to somehow use the libraries locator strategy.

Needed to also add to the test page so we had a new element after a delay.

[1] https://www.selenium.dev/selenium/docs/api/py/_modules/selenium/webdriver/common/by.html#By

emanlove committed:
Added keyword documentation for new `Wait For Expected Condition` keyword

emanlove committed:
Merge branch 'master' into add-expected-conditions-1827

yuriverweij commented:
"Wait For Expected Condition Times out within set timeout" test fails on Firefox only, strange...

yuriverweij commented on atest/acceptance/keywords/expected_conditions.robot:
I would like to add more tests to this (in the near future?)  Especially failing tests. Not sure if every expected condition should have a failing + passing test, since the handling of the expected conditions themselves is done (and tested I assume) by selenium itself. Maybe a failing+passing test for each type of expected condition (bool, webelement, list, etc.)?

emanlove replied:
Agreed

yuriverweij commented on src/SeleniumLibrary/keywords/expectedconditions.py:
```diff
@@ -0,0 +1,63 @@
+# Copyright 2016-     Robot Framework Foundation
```
This should be updated to 2024 I assume?

emanlove replied:
As I read that is starting in 2016 and going to the present the copyright belongs to the foundation.

yuriverweij reviewed (commented):
Looks good to me. I added a few comments, but I don't think they should be blocking.

emanlove committed:
Fix Firefox unit tests

Made unit test compatible to both Selenium v4.17.2 and above as well as
v4.16.0 and prior

emanlove committed:
Reduced the timeout on the expected timeout within expected conditions

The delay for the title change was 600 miliseconds while the timeout was
0.5 seconds. I suspect this gives little time for error. The 600
milisecond timeout was specifically choosen along with the other
timeouts in the expected conditions tests such that any combination of
delays would be unique. This gives us a framework to test various
combinations. So I didn't want to change that. Noting I could simply
swap the 600 millisecond test for the 900 millisecond tes .. which might
be the best bet. But for now trying to reduce the timeout to 0.3
seconds.

emanlove committed:
Expanded selenium versions upto 4.19.0

yuriverweij commented on utest/test/keywords/test_firefox_profile_parsing.py:
```diff
@@ -56,8 +56,21 @@ def test_single_method(self):
+    def _get_preferences_attribute(self, result):
+        # -- temporary fix to transition selenium to v4.17.2 from v4.16.0 and prior
+        # from inspect import signature
+        # sig = signature(result)
```
These comments can be removed I assume?

yuriverweij commented on .github/workflows/CI.yml:
```diff
@@ -11,7 +11,7 @@ jobs:
         python-version: [3.8, 3.11]  # 3.12, pypy-3.9
         rf-version: [5.0.1, 6.1.1, 7.0]
-        selenium-version: [4.14.0, 4.15.2, 4.16.0]  #4.17.0, 4.18.0
+        selenium-version: [4.14.0, 4.15.2, 4.16.0, 4.17.2, 4.18.1, 4.19.0]
```
It's becoming quite a long list again.

emanlove committed:
Fix acceptance tests

emanlove committed:
Append results to zip file

This may greatly increase the archive file size which we need to simplify our matrix.
But at least we should get some more insights into failing tests.

emanlove committed:
Added selenium version and browser name to archive

emanlove committed:
Upload only failed artifacts

emanlove committed:
Updated some expected log messages in chrome tests
